saviSimulation/savi/jason_processing/behaviour/UxVBehavior.java

```diff
@@ -0,0 +1,240 @@
+                //math to calculate angle cover
+                tan = oposite / distance;
+                angle = Math.abs(Math.atan(tan));
+
+                //if object is covered by di remove
+                //if is covered by azimuth angle
+                if (di.getParameters().get(0) + angle > visibleItems.get(i).getParameters().get(0) && di.getParameters().get(0) - angle < visibleItems.get(i).getParameters().get(0)) {
+                    //if it is covered by elevation angle
+                    if (di.getParameters().get(1) + angle > visibleItems.get(i).getParameters().get(1) && di.getParameters().get(1) - angle < visibleItems.get(i).getParameters().get(1)) {
+                        //if it's at a mayor distance
+                        if (di.getParameters().get(2) < visibleItems.get(i).getParameters().get(2)) {
+                            visibleItems.remove(visibleItems.get(i));
+                        }
+                    }
+                }
+            }
+        }
+
+        return visibleItems;
+    }
+
+    /**
+     * Adds to the current compass angle and normalizes the value between 0 and 2 * PI
+     * @param val
+     */
+    protected void addNormalizedCompassAngle(double val) {
+        this.compasAngle += val;
+        double twoPi = 2 * Math.PI;
+
+        if (this.compasAngle < 0)
+            this.compasAngle += twoPi;
+        if (this.compasAngle > twoPi)
+            this.compasAngle -= twoPi;
+    }
+
+    /**
+     * Creates the default actions for a UxV agent
+     */
+    protected void createActionMap() {
+        actionMap.put("turn(left)", (() -> this.addNormalizedCompassAngle(-Math.PI / 16.0)));
+        actionMap.put("turn(right)", (() -> this.addNormalizedCompassAngle(Math.PI / 16.0)));
+        actionMap.put("thrust(on)", (() -> this.speedVector.set(SPEED, SPEED)));
+        actionMap.put("thrust(off)", (() -> this.speedVector.set(0, 0)));
+    }
+
+    /**
+     * Process the actions in the queue, and run the corresponding action handler
+     */
+    protected void processAgentActions(List<String> actionsToExecute) {
+        for (String action : actionsToExecute) {
+            System.out.println("[ process actions] UAS id=" + this.ID + " doing: " + action);
+            AgentAction agentAction = actionMap.get(action);
+            if(agentAction == null) {
+                System.out.println("No Action defined for '" + action + "'");
+                continue;
+            }
+            agentAction.performAction();
+        }
+    }
+
+    /**
+     * Update perception Snapshot in agent state
+     */
+    protected void updatePercepts(PVector mypos) {
+        PerceptionSnapshot P = new PerceptionSnapshot();
+        //Add position
+        P.addPerception(new PositionPerception(this.time, (double) mypos.x, (double) mypos.y, (double) mypos.z));
+        //Add velocity
+        P.addPerception(new VelocityPerception(this.time, Math.atan(mypos.x / mypos.y), 0, speedVector.x));
```

Comment:
@ntartania Additional note, this line doesn't seem to be incorporating the z velocity. Should this be overridden by UAVBehaviour to provide a velocity perception that includes the z component velocity?

Reply:
I'm not sure: the thing is, Guillermo and Cristina and working on this part so I'd guess they could be implementing the 3D progressively... you might want to coordinate with them before changing any functionality...